### PR TITLE
File Uploader

### PR DESCRIPTION
**Pull request by P-Hansen**
https://github.com/bcgov/nr-bcap/issues/1326

workflow for file upload

requires [PR: 171](https://github.com/bcgov/bcgov-arches-common/pull/171) from arches common.

### CODE REVIEW

File: bcap/src/bcap/apps/Permit/components/dashboard/QuestionDialogInternal.vue
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/components/dashboard/QuestionDialogInternal.vue:1]: do we need this?

Comment (P-Hansen) [reply bcap/src/bcap/apps/Permit/components/dashboard/QuestionDialogInternal.vue:1]: oops, no we don't.

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step99_Review.vue [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    };
+}
+
+// Filters out only photographs so the generic component handles everything else
+const getFilteredFields = (fields: unknown) => {
+    if (!Array.isArray(fields)) return [];
+
+    return fields.filter((f) => {
+        const alias = f.alias || f.nodeAlias || f.node?.alias || f.node_alias;
+        return alias !== 'submission_photographs';
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step99_Review.vue]: why the different field accesses?

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step99_Review.vue [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    });
+};
+
+const getPhotos = (rawData: unknown): PhotographNode[] => {
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step99_Review.vue:51]: why unknown?

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
+import { describe, it, expect, vi } from 'vitest';
+import { defineComponent, ref } from 'vue';
+import { shallowMount } from '@vue/test-utils';
+import { useDraftStore } from '@/bcap/stores/draft.ts';
+import { submitModule } from '@/bcap/apps/Permit/api.ts';
+
+// 5. Import the Document Submission steps and module
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment, can read the line below it

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+vi.mock('@/bcap/api.ts', () => ({
+    saveDraftFieldToBackend: vi.fn(),
+}));
+
+vi.mock('@/bcap/apps/Permit/api.ts', () => ({
+    submitModule: vi.fn(),
+}));
+
+// 1. Mock the heavy GenericWidget to avoid runtime deps during shallowMount
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment, can read the line below it

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    '@/arches_component_lab/generics/GenericWidget/GenericWidget.vue',
+    () => ({
+        default: defineComponent({
+            name: 'GenericWidget',
+            template: '<div />',
+        }),
+    }),
+);
+
+// 2. Mock widget constants
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment, can read the line below it

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+vi.mock('@/arches_component_lab/widgets/constants.ts', () => ({
+    EDIT: 'edit',
+    VIEW: 'view',
+}));
+
+// 3. Mock useDraftStep composable utilized by Steps 2, 3, and 4
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment, can read the line below it

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+    Step99,
+};
+
+describe('DocumentSubmissionModule steps', () => {
+    for (const [name, component] of Object.entries(steps)) {
+        it(`${name} renders and exposes isValid()`, () => {
+            // Because we mocked the composable and store, shallowMount will succeed
+            const wrapper = shallowMount(component);
+
+            // Assert the component rendered successfully
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment, can read the line below it

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+            expect(wrapper.html()).toBeTruthy();
+
+            // Every step exposes isValid() for the stepper navigation
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment, can read the line below it

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+            expect(
+                typeof (wrapper.vm as { isValid: () => boolean }).isValid(),
+            ).toBe('boolean');
+        });
+    }
+});
+
+describe('DocumentSubmissionModule extended coverage', () => {
+    // Tests lines 2-86 in DocumentSubmissionModule.vue
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: don't need this comment will  make refactoring really painful

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+                { aliased_data: { some_prop: 'test' } },
+            ],
+            report_submission: { aliased_data: { file: 'test.pdf' } },
+            submission_photographs: [{ aliased_data: { view: 'north' } }],
+        };
+
+        const wrapper = shallowMount(DocumentSubmissionModule);
+        const stepper = wrapper.findComponent({ name: 'WorkflowStepper' });
+
+        // Trigger the custom submit function passed to the stepper
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+        await stepper.props('submit')();
+
+        // Assert the data was flattened correctly
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+                    },
+                ],
+            }),
+        );
+    });
+
+    it('Step2_Details template events', async () => {
+        const wrapper = shallowMount(Step2);
+        const widgets = wrapper.findAllComponents({ name: 'GenericWidget' });
+        // Trigger the inline updateValue() lambdas in the template
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+        for (const w of widgets) {
+            await w.vm.$emit('update:value', 'test');
+        }
+    });
+
+    it('Step3_Submission template events', async () => {
+        const wrapper = shallowMount(Step3);
+        const widgets = wrapper.findAllComponents({ name: 'GenericWidget' });
+        // Trigger the inline updateValue() lambdas in the template
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+        for (const w of widgets) {
+            await w.vm.$emit('update:value', 'test');
+        }
+    });
+
+    // Tests line 19 branch in Step 3
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: line 19, so if we refactor - you're going to have to come back and update this comment?

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+    it('Step3_Submission handles undefined draftData', () => {
+        // The global mock established at the top of the file returns an empty ref({}),
+        // which inherently tests the optional chaining fallback on line 19 safely.
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: remove

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+        const wrapper = shallowMount(Step3);
+        expect(wrapper.exists()).toBe(true);
+    });
+
+    // Tests lines 77-196 in Step 4
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: remove

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+            (w) => w.props('nodeAlias') === 'submission_photographs',
+        );
+        if (fileWidget) {
+            const mockFile = new File([''], 'test.jpg', { type: 'image/jpeg' });
+            await fileWidget.vm.$emit('update:value', {
+                node_value: [{ file: mockFile }],
+            });
+        }
+
+        // 2. Test Date Regex Formatting logic branches
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+            await dateWidget.vm.$emit('update:value', { node_value: '2020' }); // tests YYYY
+            await dateWidget.vm.$emit('update:value', {
+                node_value: '2020-01-01',
+            }); // tests YYYY-MM-DD
+            await dateWidget.vm.$emit('update:value', {
+                node_value: 'invalid',
+            }); // fallback
+        }
+
+        // 3. Test Save, Add, and Delete Photo array mutations
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+        const buttons = wrapper.findAllComponents({ name: 'Button' });
+
+        // Click Save Image
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: should be able to figure this out from reading below maybe saveBtn => saveImageBtn?

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+        const saveBtn = buttons.find((b) =>
+            (b.attributes('tooltip') || '').includes('Save'),
+        );
+        if (saveBtn) await saveBtn.vm.$emit('click');
+
+        // Click Add Image
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: same comment as above

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 21]
```diff
@@ -0,0 +1,379 @@
...
+        const addBtn = buttons.find((b) => b.attributes('label') === '+ Add');
+        if (addBtn) await addBtn.vm.$emit('click');
+
+        // Click Delete Image
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: obvious if we read the code below

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+            if (deleteIcon.exists()) await deleteIcon.trigger('click');
+        }
+
+        // Expose save method check
+        expect(
+            await (wrapper.vm as { save: () => Promise<boolean> }).save(),
+        ).toBe(true);
+    });
+
+    // Tests lines 34-35, 46, 55-56, 62 in Step 99
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: not useful for refactoring

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+    it('Step99_Review parses edge case data shapes', () => {
+        // 1. Define the stub FIRST inside the test block
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/steps.test.ts]: can read below

**Review by seeker25 — APPROVED** [review]
LGTM, I'd gun alot of your unit test comments though they are just noise and extra upkeep if refactoring

I noticed this with some of my comments in my code as well - they were stale and then the code was changed (and comment not always updated) -> which causes confusion :(

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue [hunk 24]
```diff
@@ -0,0 +1,358 @@
...
+defineExpose({ isValid: customIsValid, save: async () => true });
+</script>
+
+<template>
+    <Form
+        name="docForm"
+        :validateOnBlur="true"
+    >
+        <FieldSet legend="Document Submissions">
+            <MultiFileUploader
```
Comment (bferguso) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue:233]: Are we able to use the same pattern as the GenericWidgets here? Feels like a lot of props. It'd be nice if we could use the GenericWidget and just override the card-x-node-x-widget-data value to point at the MultiFileUploader component.

**Review by bferguso — APPROVED** [review]
Overall looks good. I think we can tune the multi file uploader a bit, but  can do that as a separate PR if you want. Let's try to keep it as close to the GenericWidget pattern as possible, only deviating where absolutely necessary.

**Review by bferguso — CHANGES_REQUESTED** [review]
Actually - let's make sure the unit test coverage is back up before we merge this in.

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue [hunk 25]
```diff
@@ -0,0 +1,385 @@
...
+    const reportData = draftData.value.report_submission?.aliased_data as
+        DocumentSubmissionReportSubmissionAliasedData | undefined;
+    if (!reportData) return false;
+
+    const titleNode = reportData.report_title;
+    const tVal = titleNode?.node_value;
+    const hasTitle = !!(
+        titleNode?.display_value ||
+        tVal?.en?.value ||
+        (typeof tVal === 'string' && (tVal as string).trim() !== '')
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue]: another `typeof`

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue [hunk 26]
```diff
@@ -0,0 +1,385 @@
...
+                        :mode="EDIT"
+                        :should-show-label="false"
+                        :aliasedNodeData="
+                            draftData?.report_submission?.aliased_data
+                                ?.archaeological_consultant
+                        "
+                        graph-slug="document_submission"
+                        node-alias="archaeological_consultant"
+                        @update:value="
+                            updateMetadata($event, 'archaeological_consultant')
```
Comment (seeker25) [inline bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue:290]: Future: ideally I think we should use some enums from aliases that way when a field changes, the alias enum breaks

File: bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue (hunk 25 above)
Comment (P-Hansen) [reply bcap/src/bcap/apps/Permit/Modules/DocumentSubmissionModule/steps/Step3_Submission.vue]: removed

**Review by bferguso — APPROVED** [review]
Looks good. I'll try to use the reload_permit_models function on the servers to see if that will pickup the model change here. 

Images shouldn't be mandatory so we'll need to fix that in the next update and we're seeing HTML on the review page, but that's a common component so we can fix that separately:

Image
